awx/main/models/inventory.py

```diff
@@ -1797,3 +1815,726 @@ class Meta:
 
     def get_absolute_url(self, request=None):
         return reverse('api:inventory_script_detail', kwargs={'pk': self.pk}, request=request)
+
+
+# TODO: move to awx/main/models/inventory/injectors.py
+class PluginFileInjector(object):
+    # if plugin_name is not given, no inventory plugin functionality exists
+    plugin_name = None  # Ansible core name used to reference plugin
+    # if initial_version is None, but we have pugin name, injection logic exists,
+    # but it is vaporware, meaning we do not use it for some reason in Ansible core
+    initial_version = None  # at what version do we switch to the plugin
+    ini_env_reference = None  # env var name that points to old ini config file
+    # base injector should be one of None, "managed", or "template"
+    # this dictates which logic to borrow from playbook injectors
+    base_injector = None
+
+    def __init__(self, ansible_version):
+        # This is InventoryOptions instance, could be source or inventory update
+        self.ansible_version = ansible_version
+
+    @property
+    def filename(self):
+        """Inventory filename for using the inventory plugin
+        This is created dynamically, but the auto plugin requires this exact naming
+        """
+        return '{0}.yml'.format(self.plugin_name)
+
+    @property
+    def script_name(self):
+        """Name of the script located in awx/plugins/inventory
+        """
+        return '{0}.py'.format(self.__class__.__name__)
+
+    def inventory_as_dict(self, inventory_update, private_data_dir):
+        """Default implementation of inventory plugin file contents.
+        There are some valid cases when all parameters can be obtained from
+        the environment variables, example "plugin: linode" is valid
+        ideally, however, some options should be filled from the inventory source data
+        """
+        if self.plugin_name is None:
+            raise NotImplementedError('At minimum the plugin name is needed for inventory plugin use.')
+        return {'plugin': self.plugin_name}
+
+    def inventory_contents(self, inventory_update, private_data_dir):
+        """Returns a string that is the content for the inventory file for the inventory plugin
+        """
+        return yaml.safe_dump(self.inventory_as_dict(inventory_update, private_data_dir), default_flow_style=False)
+
+    def should_use_plugin(self):
+        return bool(
+            self.plugin_name and self.initial_version and
+            Version(self.ansible_version) >= Version(self.initial_version)
+        )
```

Comment:
@ryanpetrello This is where I am wondering if we can add some other flag to let us use the plugin, and set the initial version to `2.9` on these things, since ansible `2.8` will be coming out soon

Reply:
I would really like to get 2.7 for some of the major sources. In particular, right now we have a valid shot at gce, we just need one image variable, and people have been making shots at this (still nothing working quite yet, but we know the general API endpoints that new requests have to be added for).

But core has a rule that features are not backported. That makes me think that 2.8 might be the furthest back we can get. That could apply to both the gce image variable and the group naming issue (or neither of them, I can't predict).

If we're going to push back any _major_ source to 2.9, we need a good reason for doing so, and resources dedicated to doing it. And I would still be rather unhappy about it.